tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics.registry;
+
+import com.google.common.collect.ForwardingSortedMap;
+import java.util.SortedMap;
+
+/**
+ * A sorted map implementation which prehashes for faster usage in hashmaps.
+ * This is only safe for immutable underlying maps.
+ */
+final class PrehashedSortedMap<K, V> extends ForwardingSortedMap<K, V> {
+    private final SortedMap<K, V> delegate;
+    private final int hashCode;
+
+    PrehashedSortedMap(SortedMap<K, V> delegate) {
```

Comment:
would suggest making immutability explicit in types

```suggestion
    PrehashedSortedMap(ImmutableSortedMap<K, V> delegate) {
```

Reply:
This would require casting in RealMetricName.

Reply:
not sure I follow that as the only caller of this constructor is `com.palantir.tritium.metrics.registry.RealMetricName#prehash` which explicitly passes in an `ImmutableSortedMap`

```
        return new PrehashedSortedMap<>(ImmutableSortedMap.copyOfSorted(map));
```

Reply:
oops!